language/extensions/move-table-extension/src/lib.rs

```diff
@@ -0,0 +1,606 @@
+        let key_bytes = serialize(&self.key_layout, key)?;
+        let key_size = key_bytes.len();
+        let mut val_size = 0;
+        if !self.content.contains_key(&key_bytes) {
+            // Try to retrieve a value from the remote resolver.
+            match context
+                .resolver
+                .resolve_table_entry(&self.handle, &key_bytes)
+                .map_err(|err| {
+                    partial_extension_error(format!("remote table resolver failure: {}", err))
+                })? {
+                Some(val_bytes) => {
+                    val_size = val_bytes.len();
+                    let val = deserialize(&self.value_layout, &val_bytes)?;
+                    self.content
+                        .entry(key_bytes.clone())
+                        .or_insert(GlobalValue::cached(val)?);
+                }
+                None => return Ok((None, key_size, val_size)),
+            }
+        }
+        Ok((
+            Some(self.content.get_mut(&key_bytes).unwrap()),
+            key_size,
+            val_size,
+        ))
+    }
+}
+
+// =========================================================================================
+// Native Function Implementations
+
+/// Returns all natives for tables.
+pub fn table_natives(table_addr: AccountAddress) -> NativeFunctionTable {
+    native_functions::make_table(
+        table_addr,
+        &[
+            ("Table", "new_table_handle", native_new_table_handle),
+            ("Table", "insert_box", native_insert_box),
+            ("Table", "length_box", native_length_box),
+            ("Table", "borrow_box", native_borrow_box),
+            ("Table", "borrow_box_mut", native_borrow_box),
+            ("Table", "remove_box", native_remove_box),
+            ("Table", "contains_box", native_contains_box),
+            ("Table", "destroy_empty_box", native_destroy_empty_box),
+            ("Table", "drop_unchecked_box", native_drop_unchecked_box),
+        ],
+    )
+}
+
+fn native_new_table_handle(
+    context: &mut NativeContext,
+    mut _ty_args: Vec<Type>,
+    args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    assert!(args.is_empty());
+
+    let table_context = context.extensions().get::<NativeTableContext>();
+    let mut table_data = table_context.table_data.borrow_mut();
+
+    // Take the transaction hash provided by the environment, combine it with the # of tables
+    // produced so far, sha256 this and select 16 bytes from the result. Given the txn hash
+    // is unique, this should create a unique and deterministic global id.
+    let mut digest = Sha3_256::new();
+    Digest::update(&mut digest, table_context.txn_hash.to_be_bytes());
+    Digest::update(&mut digest, table_data.new_tables.len().to_be_bytes());
+    let bytes: [u8; 16] = digest.finalize()[0..16].try_into().unwrap();
+    let id = u128::from_be_bytes(bytes);
+    assert!(table_data.new_tables.insert(TableHandle(id)));
```

Comment:
If multiple tables are created for the same transaction, will the generated ids be duplicated (if they all have the same length)?

Reply:
The adapter needs to provide a unique hash for the transaction (the txn_hash) and table ids would be unique relative to this hash.

Reply:
Got it, it's tables.len(), not table_data.len()